rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/TextDocumentState.java

```diff
@@ -45,64 +67,178 @@
-        return file;
+    public CompletableFuture<Versioned<List<Diagnostic>>> getCurrentDiagnosticsAsync() {
+        return getCurrentDiagnosticsAsync(Duration.ZERO);
     }
 
-    public Versioned<String> getCurrentContent() {
-        return currentContent;
+    public CompletableFuture<Versioned<List<Diagnostic>>> getCurrentDiagnosticsAsync(Duration delay) {
+        return parseAndGetCurrent(delay)
+            .thenApply(Update::getDiagnosticsAsync)
+            .thenCompose(Function.identity());
+    }
+
+    public @MonotonicNonNull Versioned<ITree> getLastTree() {
+        return last.get();
+    }
+
+    public @MonotonicNonNull Versioned<ITree> getLastTreeWithoutErrors() {
+        return lastWithoutErrors.get();
+    }
+
+    private CompletableFuture<Update> parseAndGetCurrent() {
+        var update = current;
+        update.parseIfNotParsing();
+        return CompletableFuture.completedFuture(update);
+    }
+
+    private CompletableFuture<Update> parseAndGetCurrent(Duration delay) {
+        var update = current;
+        if (update.isParsing()) {
+            return CompletableFuture.completedFuture(update);
+        } else {
+            return parseAndGetCurrentDebouncer.get(delay);
+        }
+    }
+
+    private class Update {
+        private final int version;
+        private final String content;
+        private final CompletableFuture<Versioned<ITree>> treeAsync;
+        private final CompletableFuture<Versioned<List<Diagnostic>>> diagnosticsAsync;
+        private final AtomicBoolean parsing;
+
+        public Update(int version, String content) {
+            this.version = version;
+            this.content = content;
+            this.treeAsync = new CompletableFuture<>();
+            this.diagnosticsAsync = new CompletableFuture<>();
+            this.parsing = new AtomicBoolean(false);
+        }
+
+        public Versioned<String> getContent() {
+            return new Versioned<>(version, content);
+        }
+
+        public CompletableFuture<Versioned<ITree>> getTreeAsync() {
+            parseIfNotParsing();
+            return treeAsync;
+        }
+
+        public CompletableFuture<Versioned<List<Diagnostic>>> getDiagnosticsAsync() {
+            parseIfNotParsing();
+            return diagnosticsAsync;
+        }
+
+        public boolean isParsing() {
+            return parsing.get();
+        }
+
+        private void parseIfNotParsing() {
+            if (parsing.compareAndSet(false, true)) {
+                parser
+                    .apply(location, content)
+                    .whenComplete((t, e) -> {
+                        // Prepare result values for futures
+                        var tree = new Versioned<>(version, t);
+                        var diagnostics = new Versioned<>(version, toDiagnostics(t, e));
+
+                        // Complete future to get the tree
+                        if (t == null) {
+                            treeAsync.completeExceptionally(e);
```

Comment:
loooks like we should do this before building `tree` and `diagnostics`? or are those other functions able to deal with null trees?